react/src/pages/addParticipants/AddParticipants.tsx

```diff
@@ -64,24 +64,58 @@ export default function AddParticipants() {
         }
     }
 
+    function handleDataChange(newData: DataEntryRow[]) {
+        setData(newData);
+
+        // Check required fields for all rows
+        const headers = getDataEntryHeaders();
+        let problemRows = new Map<number, Array<keyof DataEntryRowBase>>();
+
+        for (let i = 0; i < newData.length; i++) {
+            let row = newData[i];
+            for (const field of headers.required) {
+                // Condition for a row being 'problematic'
+                if (row[field].trim() === "") {
+                    if (problemRows.get(i)) problemRows.set(i, problemRows.get(i)!.concat(field));
+                    else problemRows.set(i, [field]);
+                }
+            }
+        }
+
+        if (problemRows.size > 0) {
+            let errorMessage = "Cannot submit. Required fields missing for rows:";
+            problemRows.forEach((fields, key) => {
+                const fieldStr = fields.toString().replaceAll(/\[\]/g, "");
```

Comment:
Does this show something different in your browser? I expect `Array.toString()` to behave the same as `Array.join()` so we shouldn't have any square brackets in it to begin with.

Reply:
Actually I didn't consider using `.join()` for some reason. I'll do that instead.